src/shared/util.js

```diff
@@ -1329,9 +1336,221 @@ MessageHandler.prototype = {
     return capability.promise;
   },
   /**
+   * Sends a message to the comObj to invoke the action with the supplied data.
+   * Expect that the other side will callback to signal 'start_complete'.
+   * @param {String} actionName - Action to call.
+   * @param {JSON} data - JSON data to send.
+   * @param {Object} queueingStrategy - strategy to signal backpressure based on
+   *                 internal queue.
+   * @param {Array} [transfers] - Optional list of transfers/ArrayBuffers.
+   * @return {ReadableStream} ReadableStream to read data in chunks.
+   */
+  sendWithStream(actionName, data, queueingStrategy, transfers) {
+    let streamId = this.streamId++;
+    let that = this;
+    let sourceName = this.sourceName;
+    let targetName = this.targetName;
+    return new ReadableStream({
+      start(controller) {
+        let startCapability = createPromiseCapability();
+        that.streamControllers[streamId] = {
+          controller,
+          startCall: startCapability
+        };
+        that.postMessage({
+          sourceName,
+          targetName,
+          action: actionName,
+          streamId,
+          data,
+          desiredSize: controller.desiredSize
+        });
+        // Return Promise for Async process, to signal success/failure.
+        return startCapability.promise;
+      },
+      pull(controller) {
+        let pullCapability = createPromiseCapability();
+        that.streamControllers[streamId].pullCall = pullCapability;
+        that.postMessage({
+          sourceName,
+          targetName,
+          stream: 'pull',
+          streamId,
+          desiredSize: controller.desiredSize
+        });
+        // Returning Promise will not call "pull"
+        // again until current pull is resolved.
+        return pullCapability.promise;
+      },
+      cancel(reason) {
+        let cancelCapability = createPromiseCapability();
+        that.streamControllers[streamId].cancelCall = cancelCapability;
+        that.postMessage({
+          sourceName,
+          targetName,
+          stream: 'cancel',
+          reason,
+          streamId
+        });
+        // Return Promise to signal success or failure.
+        return cancelCapability.promise;
+      }
+    }, queueingStrategy);
+  },
+
+  resolveCall(fn, args, thisArg = null) {
+    if (!fn) {
+      return Promise.resolve(undefined);
+    }
+    return new Promise((resolve, reject) => {
+      resolve(fn.apply(thisArg, args));
+    });
+  },
+
+  _createStreamSink(data) {
+    let that = this;
```

Comment:
If you're not able to avoid aliasing this (by using e.g. arrow functions below), at least use let self = this; instead since that's the convention in the code-base.